### PR TITLE
Get picture urls from API response instead of HTTP Location header

### PR DESCRIPTION
**Pull request by noahsilas**
The default behaviour for `graph.facebook.com/me/picture` and other `/picture` endpoints is to return a 302 redirect to the profile photo url. To make these `.../picture` endpoints return responses similar to the rest of the Graph API, we must add the `redirect=false` parameter [0].

Previously, in `API#get_picture` we would make the request to the graph API without this parameter, and then extract the photo URL from the 'Location' header in the HTTP response. This is a little unconventional, and may be responsible for issue #308, although I haven't tested the failure mode
in a batch query here. Instead, we can explicitly request the full API response (via `API#get_user_picture_data`), and then parse out the URL field.

I also cleaned up some of the tests and mocks to make the test suite for this particular weirdness closer to reality.

[0] The current version of the Graph API documentation doesn't seem to note this particular wart; I've opened a bug against their documentation:
https://developers.facebook.com/bugs/383635598495430/

In the meantime, you can find an old version describing this behavior thanks to the Wayback Machine:
http://web.archive.org/web/20131020151411/https://developers.facebook.com/docs/reference/api/using-pictures/

### CODE REVIEW

**Comment by noahsilas** [comment]
It's probably worth noting that `API#get_picture` is sort of a wart on the Koala public API. I would expect this method to do what `#get_user_picture_data` does. For its own part, `#get_user_picture_data` actually works on any object with a 'picture' connection, and so is also a little confusingly named.

It seems to me like `#get_picture` should be renamed to something like `get_picture_url`. Or, maybe we can deprecate this function entirely? It's not too hard to parse the url out of the response, and there's probably better error handling available to clients than replacing an error response with a `nil`.

**Comment by arsduo** [comment]
Hello! Thank you for the pull request, and sorry that it's taken me so surprisingly long to get to it. 

I think you make a good argument around `get_picture`; it's a remnant of the Facebook API's younger days and could be removed. I think I'll refactor the API to be something like what you describe and release it as part of Koala 3.0.

I'll keep you up to date with how this develops.

**Comment by arsduo** [comment]
Actually, this makes total sense. I'm merging this in now and also renaming the method `get_user_picture_data` to `get_picture_data` (with an alias).

**Comment by arsduo** [comment]
Check out #512 -- just merged it in. Those two old methods are now deprecated in favor of `get_picture_data` and will be removed in 3.0.
